### PR TITLE
Initial integration of powerdns

### PR DESCRIPTION
**Pull request by cluobah**
Direct correlation with implementing #3478. Initial attempt, feedback greatly needed and appreciated.

Modularized fog-powerdns gem lives at https://rubygems.org/gems/fog-powerdns

Advise on how to make it live at fog/powerdns instead.

Thanks!

### CODE REVIEW

**Comment by cluobah** [comment]
Oops... I'll fix this on Monday

File: lib/fog/bin/powerdns.rb [hunk 1]
```diff
@@ -0,0 +1,28 @@
+class PowerDNS < Fog::Bin
+  class << self
```
Comment (plribeiro3000) [inline lib/fog/bin/powerdns.rb]: This file should be in the provider gem.

**Comment by plribeiro3000** [comment]
Also, you should add @tokengeek, @geemus and @krames as owner in `rubygems`.

About the repo, just add @geemus as owner on it so he can move it under `fog`.

Thanks!

**Comment by geemus** [comment]
Thanks!

**Comment by cluobah** [comment]
OK I've added the owners and everything but @plribeiro3000, where should that file go in the gem? I don't have a bin directory

**Comment by plribeiro3000** [comment]
@cluobah Just create a `bin` folder inside your `lib` folder.

Take a look [here](https://github.com/fog/fog-atmos/tree/master/lib/fog/bin).

I will close this since the gem is already out there. =)

Thanks!

**Comment by geemus** [comment]
Thanks!

**Comment by plribeiro3000** [comment]
It seems this was a mistake. This PR should not be closed at all.

@cluobah Lets work towards make this happen. =)

Again, so sorry. I did a rush review and missed the context (code).

File: spec/fog/bin_spec.rb [hunk 2]
```diff
@@ -135,6 +137,7 @@
       assert_includes Fog.available_providers, "OpenStack" if OpenStack.available?
       assert_includes Fog.available_providers, "Openvz" if Openvz.available?
       assert_includes Fog.available_providers, "Ovirt" if Ovirt.available?
+      assert_includes Fog.available_providers, "ProfitBricks" if PowerDNS.available?
```
Comment (plribeiro3000) [inline spec/fog/bin_spec.rb]: Typo here, right?

Comment (cluobah) [inline spec/fog/bin_spec.rb]: Yea, I'll fix that up right now

**Comment by cluobah** [comment]
Whoa, that's weird. My VM's clock is wrong so the commits look older (3 days ago) than they are. Regardless, this is ready to merge pending your review!

**Comment by plribeiro3000** [comment]
@cluobah Nice, could you rebase it? It seems we got some merge conflicts here. =s

**Comment by cluobah** [comment]
@plribeiro3000 so it looks like the Travis build fails because [this](https://github.com/projectjellyfish/fog/blob/powerdns_integration/spec/fog/bin/powerdns_spec.rb) is looking for that file that we moved to the gem. Please advise?

File: spec/fog/bin/powerdns_spec.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+require "minitest/autorun"
+require "fog"
+require "fog/bin"
+require "helpers/bin"
+
+describe PowerDNS do
```
Comment (plribeiro3000) [inline spec/fog/bin/powerdns_spec.rb]: Shouldn't it be `Fog::PowerDNS` ?

**Comment by plribeiro3000** [comment]
@cluobah Take a look at my comment. Perhaps it is the root of the build issue.

**Comment by cluobah** [comment]
@plribeiro3000 I made that change and ran bundle exec rake travis locally, it gave me 8 other errors

**Comment by plribeiro3000** [comment]
Apparently those are the real errors since `PowerDNS` is not defined at all. Could you push this change so we can take a look at it?

**Comment by cluobah** [comment]
@plribeiro3000 done

**Comment by plribeiro3000** [comment]
@cluobah Now it seems the issue is in `fog-powerdns`. It seems to not being defining the correct constants.

**Comment by plribeiro3000** [comment]
You should define the constant `Fog::PowerDNS` [here](https://github.com/fog/fog-powerdns/blob/master/lib/fog/powerdns.rb)

**Comment by cluobah** [comment]
@plribeiro3000 So [it now](https://github.com/fog/fog-powerdns/blob/master/lib/fog/powerdns.rb) should look something like [this](https://github.com/fog/fog-atmos/blob/master/lib/fog/atmos.rb) right? My local travis build still fails with the same errors. Unless you mean something else?

**Comment by plribeiro3000** [comment]
Sure. Try to release a new version with this change. Perhaps a beta version
until we fix it for real.

And restart the build of your PR at travis. =)

LMK if you need me to do something.

Em seg, 16 de mar de 2015 às 19:19, Christopher Luo <
notifications@github.com> escreveu:

**Comment by cluobah** [comment]
@plribeiro3000 Yeah I made the change and released a new version yesterday, but I don't think I have permissions to restart build, could you do that?

**Comment by plribeiro3000** [comment]
@cluobah Just did. Which version did you release? [0.1.1](https://travis-ci.org/fog/fog/jobs/54632805#L170)?

**Comment by cluobah** [comment]
Yep

**Comment by plribeiro3000** [comment]
So it seems that was not the issue. Lets take another look at the code.

**Comment by cluobah** [comment]
How does fog react if you remove [this file](https://github.com/fog/fog/blob/master/lib/fog/bin/atmos.rb) ?

**Comment by plribeiro3000** [comment]
Lets see. Just a minute. =)

**Comment by plribeiro3000** [comment]
Take a look at #3489.

**Comment by plribeiro3000** [comment]
@cluobah It seems to be ok without that file. All tests passing. So it gotta be something in the modular provider. Give me some time to review the code in `fog-powerdns`.

**Comment by cluobah** [comment]
@plribeiro3000 no problem, thanks for your help!

**Comment by plribeiro3000** [comment]
@cluobah Got it. You gotta do a `require` in this [file](https://github.com/fog/fog/blob/master/lib/fog.rb) inside `Fog`.

**Comment by cluobah** [comment]
@plribeiro3000 Are you sure? I have [that](https://github.com/projectjellyfish/fog/blob/powerdns_integration/lib/fog.rb) in my PR

**Comment by plribeiro3000** [comment]
Sorry. wrong link. Try [this one](https://github.com/fog/fog/blob/master/lib/fog/bin.rb)

**Comment by cluobah** [comment]
Let's see!

**Comment by plribeiro3000** [comment]
@cluobah Done. Could you rebase those commits?

**Comment by cluobah** [comment]
@plribeiro3000 woot

**Comment by plribeiro3000** [comment]
@cluobah Thank you!

**Comment by geemus** [comment]
Thanks!

**Pull request merged**